clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleHttpLoadBalancerAtomicOperation.groovy

```diff
@@ -0,0 +1,450 @@
+    UrlMap existingUrlMap = null
+    try {
+      existingUrlMap = compute.urlMaps().get(project, urlMapName).execute()
+    } catch (GoogleJsonResponseException e) {
+      // 404 is thrown if the url map doesn't exist. Any other exception needs to be propagated.
+      if (e.getStatusCode() != 404) {
+        throw e
+      }
+    }
+
+    // Determine if the infrastructure in the description exists already.
+    // If it does, check and see if we need to update it from the description.
+
+    // UrlMap
+    urlMapExists = existingUrlMap as Boolean
+    urlMapNeedsUpdated = GCEUtil.shouldUpdateUrlMap(existingUrlMap, description)
+
+    // ForwardingRule
+    ForwardingRule existingRule = null
+    try {
+      existingRule = compute.globalForwardingRules().get(project, httpLoadBalancerName).execute()
+    } catch (GoogleJsonResponseException e) {
+      if (e.getStatusCode() != 404) {
+        throw e
+      }
+    }
+    forwardingRuleExists = existingRule as Boolean
+    if (forwardingRuleExists) {
+      forwardingRuleNeedsUpdated = (httpLoadBalancer.ipAddress ? existingRule.getIPAddress() != httpLoadBalancer.ipAddress : false) ||
+        (httpLoadBalancer.ipProtocol ? existingRule.getIPProtocol() != httpLoadBalancer.ipProtocol : false) ||
+        (description.certificate ? false : existingRule.getPortRange() != "${httpLoadBalancer.portRange}-${httpLoadBalancer.portRange}")
+    }
+
+    // TargetProxy
+    def existingProxy = null
+    if (forwardingRuleExists) {
+      String targetProxyType = Utils.getTargetProxyType(existingRule.getTarget())
+      String targetProxyName = GCEUtil.getLocalName(existingRule.getTarget())
+      switch (targetProxyType) {
+        case "targetHttpProxies":
+          existingProxy = compute.targetHttpProxies().get(project, targetProxyName).execute()
+          targetProxyNeedsUpdated = httpLoadBalancer.certificate as Boolean
+          forwardingRuleNeedsUpdated = forwardingRuleNeedsUpdated || targetProxyNeedsUpdated // We need to change the target of the forwarding rule if we change proxy from Http -> Https.
+          break
+        case "targetHttpsProxies":
+          existingProxy = compute.targetHttpsProxies().get(project, targetProxyName).execute()
+          targetProxyNeedsUpdated = httpLoadBalancer.certificate ?
+            GCEUtil.getLocalName(existingProxy?.getSslCertificates()[0]) != GCEUtil.getLocalName(GCEUtil.buildCertificateUrl(project, httpLoadBalancer.certificate)) :
+            false // We shouldn't downgrade the proxy if a cert is absent in the request.
+          break
+        default:
+          log.warn("Unexpected target proxy type for $targetProxyName.")
+          break
+      }
+      targetProxyExists = existingProxy as Boolean
+    }
+
+    // HttpHealthChecks
+    // I don't know why, but if I compare the RHS of the next two statements directly the comparison doesn't work.
```

Comment:
Try to assign them static types and see what happens. Or print out the classname of each? Or use debugger.